### ISSUE TITLE
Adapter from Repository to fs.FS

### ISSUE THREAD

**Issue by charlesdaniels** [fyne-io/fyne#2680]
Starting in Go 1.16, the [`io/fs`](https://pkg.go.dev/io/fs) package provides a set of interfaces for abstracting filesystem access. Where possible we should inter-operate with these types with the storage repository system. It will be a bit of an odd fit though - storage repositories are designed to handle arbitrary URIs, while `io/fs` assume you have an explicit filesystem root.

My instinct would be to create two adapters. One would adapt a storage repository to an `fs.FS` rooted at a particular URI, something like `NewFSFromURI(uri fyne.URI)`. Nearly all of the methods translate directly or almost directly, except for stat-related things (see also: #2233, #2134). The other adapter would go the other way around, creating a repository from an `fs.FS`, probably something like `RegisterFSAsRepository(protocol string)`. The latter may be the impetus for finally implementing the "routed repository" (a repository which registers for a protocol as usual, but internally routes access to specific URIs to nested sub-repositories).

These two things are probably separate PRs.

**NOTE**: we still support back to Go 1.14. Until 1.16 is our minimum supported version, we can't implement this in a way that makes the build depend on it. 

- [ ] `fs.FS` -> repository
    - [ ] proposal (insert link here)
    - [ ] PR (insert link here)
- [ ] repository -> `fs.FS`
    - [ ] proposal (insert link here)
    - [ ] PR (insert link here)

**Comment by Jacalz** [comment]
This could, in theory, be protected by the `go116` build-tags to only make the API available on Go 1.16 and above. I don't know how good this is from a developer point, but it is a possibility.

**Comment by changkun** [comment]
(Despite not possible, everything would be much easier if we bump the Go version to 1.16 :)

**Comment by andydotxyz** [comment]
> This could, in theory, be protected by the `go116` build-tags to only make the API available on Go 1.16 and above. I don't know how good this is from a developer point, but it is a possibility.

I don’t think we should have certain APIs only available on some Go versions.

**Comment by bon-ami** [comment]
I'm using gin, which needs http.FileSystem. Since not supported by Fyne, I have to begin working on it myself... :-(

**Comment by andydotxyz** [comment]
The core library still supports down to ago v1.14 which is why we have not been able to work on this yet.

**Comment by changkun** [comment]
Do we need to support 1.14 when 1.20 is out? Or when would be the best point to bump to 1.16 and 1.18?

**Comment by andydotxyz** [comment]
The latest release that is generally available in all package systems is 1.15 (once Debian releases next version we can review).

Also the nice thing about 1.14 is we still support iOS 32bit apps. Once Debian updates we can look seriously at this again and see what version is well supported everywhere. And what has compelling features to justify the upgrade. I'm expecting 1.16 becomes our baseline after Debian 12 is released... if we can wait that long.

**Comment by Jacalz** [comment]
The next version of Debian (Bookworm) seem to have https://packages.debian.org/bookworm/golang-go Go 1.19 at the moment and is planning to release this summer as far as I know. While I don't think we would want to jump as far as 1.19, we should consider carefully which version we want to as the next minimum before we commit to anything.

**Comment by Jacalz** [comment]
FYI: For anyone wanting to look at this, we support a new enough Go version for this to be implemented :)
